Workspace awareness

andrewdavidmackenzie opened the pull request:
Here is a WIP Pull-request that implements simple workspace awareness.

Care to take a look at give me some review comments / feedback?

I have left in some comments that go beyond the scope of the actual code changes as FYI.
I would take parts of them and make them separate enhancement issues for cargo-bundle if you agree, and remove the comments - so the comments match the code.

Note that I extracted CargoSettings from Settings (and used just Package) as:
- only the Package part of CargoSettings was actually used
- the CargoSettings implementation is now more alligned with all possible Cargo.toml files (where package is optional), so it can also be used for reading parent Cargo.toml files in a workspace

andrewdavidmackenzie commented:
Any comments?

mdsteele commented on src/bundle/settings.rs:
```diff
@@ -44,15 +44,22 @@ struct PackageSettings {
-    package_type: Option<PackageType>, // If `None`, use the default package type for this os
+    package: PackageSettings,
+    package_type: Option<PackageType>,
+    // If `None`, use the default package type for this os
```
It's a little confusing to have this comment here, since it looks like it applies to the `target` field.  If it won't fit on the end of the `package_type` line, then could we move it to just above that line?

andrewdavidmackenzie replied:
Hmm... seems like it's the IntelliJ automatic formatting that pushed it to a new line
(a new line that doesn't make sense IMO).... so will put it back manually.

mdsteele commented on src/bundle/settings.rs:
```diff
@@ -100,23 +99,90 @@ impl Settings {
+    /*
+        Try to load a set of CargoSettings from a "Cargo.toml" file in the specified directory
+    */
+    fn cargo_settings(dir: &PathBuf) -> ::Result<CargoSettings> {
```
Seems like this would make more sense as a method of `CargoSettings`, rather than `Settings`.  Maybe `CargoSettings::load()`?

andrewdavidmackenzie replied:
Yes, off course! Makes sense.

mdsteele commented on src/bundle/settings.rs:
```diff
@@ -76,22 +83,14 @@ impl Settings {
+        let cargo_settings = Settings::cargo_settings(&current_dir)?;
+        let package = match cargo_settings.package {
+            Some(package_info) => package_info,
+            None => return Err("No 'package' info found in 'Cargo.toml'".into())
```
In this case, this can be written more succinctly as just:
```
None => bail!("No 'package' info found in 'Cargo.toml'"),
```

andrewdavidmackenzie replied:
I was getting some "note: this error originates in a macro outside of the current crate
" errors, but will try to fix them.

mdsteele commented on src/bundle/settings.rs:
```diff
@@ -100,23 +99,90 @@ impl Settings {
+             to, or do we have to walk up the file system until "/" to see if we can find a
+             Cargo.toml?
+    */
+    fn get_workspace_dir(current_dir: &PathBuf) -> PathBuf {
```
Looks like this method is still WIP?

andrewdavidmackenzie replied:
This is where I mentioned that the comments are more on a "full" implementation, but the supplied implementation is minimal.....so I propose to strip the comments back to match the code, and add issues about future enhancements - if you agree.

andrewdavidmackenzie commented:
OK, I addressed those comments - take a look.
I'll create those ER issues I mention, that would build on top of this to improve how some settings would be heeded, including workspace related ones.

mdsteele commented:
Thanks for the changes!
